utils/src/maybe_encrypted.rs

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::marker::PhantomData;
+
+use crypto::{random::make_true_rng, symkey::SymmetricKey};
+use serialization::{Decode, DecodeAll, Encode};
+use zeroize::Zeroizing;
+
+use thiserror::Error;
+
+#[derive(Debug, PartialEq, Eq, Clone, Error)]
+pub enum MaybeEncryptedError {
+    #[error("Error in decryption: {0}")]
+    DecryptionError(#[from] crypto::symkey::Error),
+
+    #[error("Error in decoding: {0}")]
+    DecodingError(#[from] serialization::Error),
+}
+
+/// A generic type that stores a value as a vector of bytes that are optionally encrypted
+#[derive(Encode, Decode)]
+pub struct MaybeEncrypted<T: Encode + Decode + DecodeAll> {
+    value: Vec<u8>,
+    _phantom: PhantomData<fn() -> T>,
+}
+
+impl<T: Decode + Encode + DecodeAll> MaybeEncrypted<T> {
```

Comment:
DecodeAll is implied when Decode is implemented. You can see the impl block of DecodeAll imposing Decode trait bound.

Reply:
removed Decode